test/GPv2Encoding.test.ts

```diff
@@ -327,6 +329,78 @@ describe("GPv2Encoding", () => {
         .be.reverted;
     });
 
+    it("should verify EIP-1271 contract signatures by returning owner", async () => {
+      const artifact = await artifacts.readArtifact("ERC1271Verifier");
+      const verifier = await waffle.deployMockContract(deployer, artifact.abi);
+
+      const message = eip1271Message(testDomain, sampleOrder);
+      const eip1271Signature = "0x031337";
+      await verifier.mock.isValidSignature
+        .withArgs(message, eip1271Signature)
+        .returns(EIP1271_MAGICVALUE);
+
+      const encoder = new SettlementEncoder(testDomain);
+      await encoder.encodeContractTrade(
+        sampleOrder,
+        verifier.address,
+        eip1271Signature,
+      );
+
+      const { trades } = await encoding.decodeTradesTest(
+        encoder.tokens,
+        encoder.encodedTrades,
+      );
+
+      expect(trades.length).to.equal(1);
+      const { owner } = decodeTrade(trades[0]);
+      expect(owner).to.equal(verifier.address);
+    });
+
+    it("should revert on an invalid EIP-1271 signature", async () => {
+      const message = eip1271Message(testDomain, sampleOrder);
+      const eip1271Signature = "0x031337";
+
+      const artifact = await artifacts.readArtifact("ERC1271Verifier");
+      const verifier1 = await waffle.deployMockContract(deployer, artifact.abi);
+
+      await verifier1.mock.isValidSignature
+        .withArgs(message, eip1271Signature)
+        .returns(EIP1271_MAGICVALUE - 1);
+
+      const encoder1 = new SettlementEncoder(testDomain);
+      await encoder1.encodeContractTrade(
+        sampleOrder,
+        verifier1.address,
+        eip1271Signature,
+      );
+      await expect(
+        encoding.decodeTradesTest(encoder1.tokens, encoder1.encodedTrades),
+      ).to.be.revertedWith("invalid eip1271 signature");
+
+      const NON_STANDARD_ERC1271_VERIFIER = [
+        "function isValidSignature(bytes32 _hash, bytes memory _signature)",
+      ]; // no return value
+      const verifier2 = await waffle.deployMockContract(
+        deployer,
+        NON_STANDARD_ERC1271_VERIFIER,
+      );
+
+      await verifier2.mock.isValidSignature
+        .withArgs(message, eip1271Signature)
+        .returns();
+
+      const encoder2 = new SettlementEncoder(testDomain);
+      await encoder2.encodeContractTrade(
+        sampleOrder,
+        verifier2.address,
+        eip1271Signature,
+      );
+      // Transaction reverted: function returned an unexpected amount of data
+      await expect(
+        encoding.decodeTradesTest(encoder2.tokens, encoder2.encodedTrades),
+      ).to.be.reverted;
+    });
```

Comment:
Could you add a test case that verifies we fail if `isValidSignature` is trying to modify state (as in it's implementation is not view)?

Another classic test would be that it fails when applied to an EOA (cf. the [0x bug](https://blog.0xproject.com/post-mortem-0x-v2-0-exchange-vulnerability-763015399578) back in the day).

Reply:
Great test cases and nice read!